dv/cosim/spike_cosim.cc

```diff
@@ -68,40 +68,59 @@ SpikeCosim::SpikeCosim(const std::string &isa_string, uint32_t start_pc,
-  bool bus_error = !bus.load(addr, len, bytes);
+  // Load from the ISS memory model, while checking if the access is
+  // consistent with the observed DUT memory acceses.
+  // Return false if any errors occur, or the checking fails.
+  // Within Spike (mmu_t::load_slow_path), this causes a load_access_fault.
+
+  // Spike does not differentiate between iside and dside memory accesses
+  // (as ibex does) so we make an educated guess which is which, and
+  // then after fetching the data, check against the recorded DUT access
+  // that should be queued as pending.
+  //
+  // Spike may attempt to access up to 8-bytes from the PC when fetching, so
+  // assume as a dside access when it falls outside that range.
+  // N.B. Heuristic
+  uint32_t pc = processor->get_state()->pc;
+  bool is_probably_dmem_access = (addr < pc || addr >= (pc + 8));
 
-  bool dut_error = false;
+  // Attempt to make an identical load from the bus memory device.
+  // This can fail e.g. if accessing an unmapped address (see SpikeCosim::add_memory)
+  bool bus_error = !bus.load(addr, len, bytes);
 
-  // Incoming access may be an iside or dside access. Use PC to help determine
-  // which.
-  uint32_t pc = processor->get_state()->pc;
-  uint32_t aligned_addr = addr & 0xfffffffc;
+  // If the DUT encountered a bus error for its access, or the checking failed,
+  // this is a dut_error.
+  bool dut_error = is_probably_dmem_access ? (check_mem_access(false, addr, len, bytes) != kCheckMemOk) : false;
 
-  if (pending_iside_error && (aligned_addr == pending_iside_err_addr)) {
-    // Check if the incoming access is subject to an iside error, in which case
-    // assume it's an iside access and produce an error.
-    pending_iside_error = false;
-    dut_error = true;
-  } else if (addr < pc || addr >= (pc + 8)) {
-    // Spike may attempt to access up to 8-bytes from the PC when fetching, so
-    // only check as a dside access when it falls outside that range.
+  if (!is_probably_dmem_access) {
+    // This is probably an iside access.
 
-    // Otherwise check if the aligned PC matches with the aligned address or an
+    // Check if the aligned PC matches with the aligned address or an
     // incremented aligned PC (to capture the unaligned 4-byte instruction
-    // case). Assume a successful iside access if either of these checks are
-    // true, otherwise assume a dside access and check against DUT dside
-    // accesses.  If the RTL produced a bus error for the access, or the
-    // checking failed produce a memory fault in spike.
-    dut_error = (check_mem_access(false, addr, len, bytes) != kCheckMemOk);
-  }
+    // case).
+    uint32_t aligned_addr = addr & 0xfffffffc;
+    if ( pending_iside_error &&
+        (pending_iside_err_addr == aligned_addr)) {
+      // If the pending_iside_error has been set, produce a dut_error.
+      pending_iside_error = false;
+      dut_error = true;
+    } else {
+      // Probably a valid iside access
```

Comment:
I can see the argument documentation wise but we shouldn't really have an empty else, so ditch it and move the 'Probably a valid iside access' to an appropriate point.